### ISSUE TITLE
Make causal mask in TransformerDecoder optional

### ISSUE THREAD

**Issue by ferraric** [keras-team/keras-nlp#1062]
**Is your feature request related to a problem? Please describe.**
The [TransformerDecoder](https://keras.io/api/keras_nlp/modeling_layers/transformer_decoder/) layer in keras-nlp will always apply a "causal-mask" (mask that prevents tokens to attending to tokens that come after it in the sequence order) to the decoder attention layer.
Currently there is no way of using this layer without that attention constraint.

**Describe the solution you'd like**
I would like to be able to use a TransformerDecoder layer without the causal mask.

**Describe alternatives you've considered**
I could implement the TransformerDecoder layer myself.

**Additional context**
My use case is that I have a classification task of a sequence A where I need to classify every element of the sequence. In addition, I have a sequence B that adds general context. In an architecture similar to Attention is all you need, I process A as the decoder sequence and B as the encoder sequence. In this use case I do not want to restrict attention between elements of A because they are all classified at once.

Since the TransformerDecoder layer is part of the public API, I believe it to be a worthy feature for it to be useable without the causal mask. Even though I understand, the common use case is for this to be employed as part of an autoregressive model where the causal mask is essential, making the mask optional (with of course the default still being True) is a natural generalization and would make the layer more generally useful.

**Comment by jbischof** [comment]
@ferraric thanks for the suggestion! Isn't the causal mask the only thing differentiating the decoder block from an encoder block? Why not use two encoder blocks instead?

**Comment by ferraric** [comment]
@jbischof thanks for the prompt response. The causal mask is not the only thing differentiating the decoder block from the encoder block. The crucial difference is the cross attention between the encoder and the decoder sequence (Encoder-Decoder Attention in the illustration below). Since the interface to the encoder block only accepts one sequence, I don't see how one could mimic the desired behavior with just encoder blocks.
If I am missing something, let me know

Encoder block:
![image](https://github.com/keras-team/keras-nlp/assets/32451307/64ee5a01-e1ca-4ac5-b14c-44ee206aa9c2)

Decoder block:
![image](https://github.com/keras-team/keras-nlp/assets/32451307/310b56a2-9c93-48d2-a723-bb9bd7627fa9)

source: http://jalammar.github.io/illustrated-transformer/

**Comment by jbischof** [comment]
@ferraric do you want to contribute a PR?

**Comment by jbischof** [comment]
No, we just want to unblock you if possible

**Comment by ferraric** [comment]
Not sure about your notification settings, therefore commenting here that I opened a [PR](https://github.com/keras-team/keras-nlp/pull/1083) for this
